### ISSUE TITLE
Electron Auto Updater MacOS support

### ISSUE THREAD

**Issue by badams** [electron-userland/electron-builder#1089]
My intention here is to start a discussion about how to implement auto-updater support for MacOS and eventually Linux.

What changes need to be done in order for this work?

1. **latest.yml changes**
I would suggest adapting a new format similar to [this](https://github.com/megahertz/electron-simple-updater/blob/master/example/updates.json). The generation of this file would also need to be moved to outside of the [targets/nsis.ts](https://github.com/electron-userland/electron-builder/blob/master/packages/electron-builder/src/targets/nsis.ts#L253-L261) to somewhere more appropriate for handling multiple platforms.

3. **Actual update process**
I dont see any reason for the check-updates/download process to change much, the downloading should work fine across platforms, and app-update.yml should only contain details for the current platform. _Actually installing the update is something I'm not clear about how that is achieved on MacOS?_ perhaps a shell script that replaces the contents of the Applications/MyApp with the zip contents?

Any other points/suggestions?

**Comment by develar** [comment]
Solution from https://github.com/megahertz/electron-simple-updater will be merged into electron-auto-updater.

>  I'm not clear about how that is achieved on MacOS? perhaps a shell script that replaces the contents of the Applications/MyApp with the zip contents?

Squirrel.Mac will be not replaced. "Hack" to avoid extra server just applied.

When it will be done? I hope this weekend.

**Comment by develar** [comment]
@badams If you can help to merge macOS solution from https://github.com/megahertz/electron-simple-updater — it will be cool.

**Comment by develar** [comment]
Ok — I started to work on it.

**Comment by develar** [comment]
For macOS postfix `-mac` will be used. i.e. in addition to `latest.yml` for windows, file `latest-mac.yml` will be generated (here `latest` it is channel name (supported by generic provider)).

**Comment by badams** [comment]
@develar sorry I haven't had a lot of time to get on github - Thanks for putting this together! :)

**Comment by amilajack** [comment]
Do windows and linux allow for serverless updates?

**Comment by develar** [comment]
@amilajack Yes, on windows you can use Github, Bintray or any generic http server (e.g. amazon s3, https included). On Linux auto-update not yet implemented.

**Comment by farfromrefug** [comment]
@develar I am trying to test your dev on this.
I am using generic as a publish provider. How will it work? i don't see any latest.yml being generated.

Thanks

**Comment by develar** [comment]
@farfromrefug If you brave enough, you can test, but don't expect that it will work.

> How will it work? i don't see any latest.yml being generated.

We don't generate latest.yml — `latest-mac.json` is generated. Why JSON? Because this file is used as release.json file for Squirrel.Mac.

Also, please note — electron-auto-updater is not yet published — you need to use it from source (or link / or publish for you).

All required changes on electron-builder is available in the electron-builder 11.2.5.

**Comment by farfromrefug** [comment]
@develar sure i am willing :D and i know it's in dev. I though a mentin about that json file, but i don't see it either. I have gone through the macPackager.ts but i don't see any mention of it

Will switch to github builds for electron-builder and electron-auto-updater, and see what i can find

Thanks

**Comment by develar** [comment]
>  I have gone through the macPackager.ts but i don't see any mention of it

https://github.com/electron-userland/electron-builder/blob/master/packages/electron-builder/src/publish/PublishManager.ts

@farfromrefug Do you build dmg? File will be generated only if `zip` target presented (default target for mac is `dmg` + `zip`.

**Comment by farfromrefug** [comment]
Ok that s what i thought, but the publishManager is not created using generic.
Yes i am building a zip

**Comment by develar** [comment]
@farfromrefug Ouch, I got what do you mean. Well, please try to use `11.2.4` (should generates the same file). Thank you! It is "regression" in the latest release.

**Comment by farfromrefug** [comment]
ok will do. ```--publish=always``` won't help as you check for authentication after that

**Comment by develar** [comment]
@farfromrefug `latest-mac.json` generation fixed in 11.2.6.

**Comment by develar** [comment]
It works. Really. No more special server for macOS updates. It will be released in 2 days (depends on my free time), because auto-update on windows should be checked manually (to avoid any regressions).

**Comment by badams** [comment]
Awesome work @develar! I will do some testing too when I get time

**Comment by develar** [comment]
Miracles happens. Serverless macOS auto-update is implemented in the electron-builder 11.3.0 and electron-auto-updater 1.0.0 Enjoy. It is not yet polished, but works.

Thanks to @megahertz for the idea.

**Comment by megahertz** [comment]
@develar Actually it was not my idea. But Unfortunately I don't remember where I've found it.

**Comment by farfromrefug** [comment]
@develar there is an issue with the name of zip file created while packaging and the zip path contained in the latest-mac.json.
[in the json](https://github.com/electron-userland/electron-builder/blob/2252964cb3c03fee565e7dfb6ebf29fce67c42fa/packages/electron-builder/src/publish/PublishManager.ts#L196) you use production name
while as [in packaging](https://github.com/electron-userland/electron-builder/blob/88682e826c36d5032df6fe42c45aa60f9739da35/packages/electron-builder/src/targets/ArchiveTarget.ts#L24) you use dev name.

In my case ```productFilename``` and ```name``` are not the same (but i think it is the case for any app)

**Comment by develar** [comment]
@farfromrefug It is because you use generic provider. Yes, due to silly GitHub restrictions and Squirrel.Mac bug (cannot use relative path) we must generate special `latest-mac.json` for GitHub.

**Comment by farfromrefug** [comment]
@develar ok so what's the solution here? Wouldn't it be a good solution to name the zip with ```name``` instead of ```productName```?

**Comment by develar** [comment]
@farfromrefug Solution is to generate file for GitHub in the `github` directory.

> Wouldn't it be a good solution to name the zip with name instead of productName?

Yes, it is another variant. But sometimes user expect product name here as well.

PR is welcome or I will address it in 2 days somehow.

**Comment by farfromrefug** [comment]
@develar i totally understand. And i guess you are right. People do expect productName. Even i expect it ;)
I think the best solution for now is to rename the file afterwards. It's really easy for me as i use gulp with your API

Will report more on the whole update process. Thanks for your amazing work!

**Comment by develar** [comment]
@farfromrefug Fixed in 11.4.2 — GitHub file in the `github` directory.

**Comment by develar** [comment]
`electron-auto-updater` renamed to `electron-updater`
